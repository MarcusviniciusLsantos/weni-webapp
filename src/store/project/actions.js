import { captureException } from '@sentry/browser';
import projects from '../../api/projects';
import i18n from '../../utils/plugins/i18n';
import { unnnicCallAlert } from '@weni/unnnic-system';
import { fetchFlowOrganization } from '../org/actions';

export default {
  getCurrentProject() {
    const object = window.localStorage.getItem('project');
    if (object) return JSON.parse(object);
    return null;
  },

  getProject(store, { uuid }) {
    return projects.getProject({ uuid });
  },

  getProjects(store, { orgId, page = 1, limit = 20, ordering }) {
    const offset = limit * (page - 1);
    return projects.list(orgId, offset, limit, ordering);
  },

  async createProjectForOrg(
    {
      commit,
      rootState: {
        Org: {
          currentOrg: { uuid },
        },
      },
    },
    { project, onBack, onAccess },
  ) {
    commit('PROJECT_CREATE_REQUEST');
    try {
      let response = null;

      response = await projects.createReadyMadeProject(
        project.name,
        uuid,
        project.dateFormat,
        project.timeZone,
        project.format,
        project.globals,
      );

      let flowOrganization = response.data.flow_organization;

      if (!flowOrganization) {
        flowOrganization = await fetchFlowOrganization(response.data.uuid);
      }

      commit('PROJECT_CREATE_SUCCESS', {
        ...response.data,
        flow_organization: flowOrganization,
      });

      commit('OPEN_MODAL', {
        type: 'confirm',
        data: {
          icon: 'check-circle-1-1',
          scheme: 'feedback-green',
          title: i18n.t('projects.create.confirm_title'),
          description: i18n.t('projects.create.confirm_subtitle'),
          cancelText: i18n.t('projects.create.view_projects'),
          confirmText: i18n.t('projects.create.go_to_project'),
          onClose: onBack,
          onConfirm: (justClose) => {
            justClose();
            onAccess(response.data.uuid);
          },
        },
      });
    } catch (e) {
      commit('PROJECT_CREATE_ERROR', e);
      unnnicCallAlert({
        props: {
          text: i18n.t('projects.create.error'),
          title: 'Error',
          icon: 'check-circle-1-1',
          scheme: 'feedback-red',
          position: 'bottom-right',
          closeText: i18n.t('close'),
        },
        seconds: 3,
      });

      captureException(new Error('PROJECT_CREATE', { cause: e }));
    }
  },

  async changeReadyMadeProjectProperties(
    { getters, commit },
    { projectUuid, first_access },
  ) {
    const { data } = await projects.changeReadyMadeProjectProperties({
      projectUuid,
      first_access,
    });

    commit('setCurrentProject', {
      ...getters.currentProject,
      first_access: data.first_access,
    });
  },

  editProject(store, { name, organization, projectUuid, timezone }) {
    return projects.editProject(name, organization, projectUuid, timezone);
  },

  deleteProject(store, { uuid }) {
    return projects.deleteProject(uuid);
  },

  setCurrentProject(
    { commit },
    {
      uuid,
      name,
      menu = {
        chat: [],
        flows: '',
        inteligence: '',
      },
      organization = {
        uuid: '',
      },
      flow_organization = {
        uuid: '',
        id: '',
      },
      first_access,
      flow_uuid,
      flow_count,
      project_type,
      authorization,
    } = {},
  ) {
    commit('setCurrentProject', {
      uuid,
      name,
      menu,
      organization,
      flow_organization,
      first_access,
      flow_uuid,
      flow_count,
      project_type,
      redirect_url: '',
      authorization,
    });
  },

  clearCurrentProject({ commit }) {
    commit('setCurrentProject', null);
  },

  async createOrUpdateProjectAuthorization(
    store,
    { email, projectUuid, role, chatRole, hasChat },
  ) {
    return projects.createProjectAuthorization({
      email,
      projectUuid,
      role,
      chatRole,
      hasChat,
    });
  },

  async removeProjectAuthorization(store, { email, projectUuid }) {
    return projects.deleteProjectAuthorization({
      email,
      projectUuid,
    });
  },

  async getSuccessOrgStatusByFlowUuid({ state, commit }, { flowUuid, force }) {
    try {
      if (!state.championChatbots[flowUuid] || force) {
        const { has_ia, has_flows, has_channel, has_msg } =
          await projects.apiFlowsGetSuccessOrg({ flowUuid });

        commit('setChampionChatbot', {
          flowUuid,
          has_ia,
          has_flows,
          has_channel,
          has_msg,
        });
      }

      return state.championChatbots[flowUuid];
    } catch (error) {
      commit('setChampionChatbot', {
        flowUuid,
        error: true,
      });

      throw error;
    }
  },
};
