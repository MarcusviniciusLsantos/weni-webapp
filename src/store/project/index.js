import { reactive } from 'vue';
import projects from '../../api/projects';
import actions from './actions';
import mutations from './mutations';

const recentActivities = {};

const state = {
  currentProject: null,
  loadingCreateProject: false,
  errorCreateProject: null,
  templates: {
    status: null,
    data: [],
  },
  championChatbots: {},
  projects: [],
  recentActivities: new Proxy(recentActivities, {
    get(target, name, receiver) {
      if (!Reflect.has(target, name)) {
        const recentActivities = reactive({
          status: null,
          next: null,
          data: [],
        });

        recentActivities.loadNext = async () => {
          recentActivities.status = 'loading';

          await projects
            .latestActivities({
              projectUuid: name,
              limit: 20,
              next: recentActivities.next,
            })
            .then(({ data }) => {
              if (data.next) {
                const url = new URL(data.next);
                const cursor = url.searchParams.get('cursor');
                recentActivities.next = cursor;
              }

              const results = [...recentActivities.data, ...data.results];

              recentActivities.data = results.filter(
                (value, index, self) =>
                  index ===
                  self.findIndex(
                    (item) =>
                      item.user === value.user &&
                      item.created_at === value.created_at,
                  ),
              );

              recentActivities.status = data.next == null ? 'complete' : null;
            })
            .catch(() => {
              console.log('error');
              if (recentActivities.status !== 'complete') {
                recentActivities.status = 'error';
              }
            });
        };

        Reflect.set(target, name, recentActivities, receiver);
      }

      return Reflect.get(target, name, receiver);
    },
  }),
};

const getters = {
  currentProject(state) {
    return state.currentProject;
  },
};

export default {
  state,
  getters,
  actions,
  mutations,
};
