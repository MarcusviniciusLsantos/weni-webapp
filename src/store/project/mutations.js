import Vue from 'vue';
import store from '../../store';

export default {
  setCurrentProject: (state, project) => (state.currentProject = project),

  setChampionChatbot: (state, value) => {
    Vue.set(state, 'championChatbots', {
      ...state.setChampionChatbots,
      [value.flowUuid]: value,
    });
  },

  PROJECT_CREATE_REQUEST: (state) => (state.loadingCreateProject = true),
  PROJECT_CREATE_SUCCESS: (state, project) => {
    const projects = store.state.Project.projects.find(
      ({ orgUuid }) => orgUuid === project.organization,
    );

    if (projects) {
      projects.data.push(project);
    }

    state.currentProject = project;
    state.loadingCreateProject = false;
  },
  PROJECT_CREATE_ERROR: (state, createProjectError) => {
    state.errorCreateProject = createProjectError;
    state.loadingCreateProject = false;
  },
};
