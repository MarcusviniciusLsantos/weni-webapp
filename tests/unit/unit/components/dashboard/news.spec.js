import { vi } from 'vitest';
import { shallowMount, createLocalVue, RouterLinkStub } from '@vue/test-utils';
import Vuex from 'vuex';

import news from '@/components/dashboard/news.vue';
import i18n from '@/utils/plugins/i18n';

import project from '../../../__mocks__/project';

const localVue = createLocalVue();
localVue.use(Vuex);

describe('news.vue', () => {
  let wrapper;
  let store;
  let getters;

  beforeEach(() => {
    getters = {
      currentProject() {
        return project;
      },
    };
    store = new Vuex.Store({
      getters,
    });
    wrapper = shallowMount(news, {
      localVue,
      store,
      i18n,
      mocks: {
        $t: () => 'some specific text',
        setTimeout: function () {
          return 99;
        },
        clearTimeout: function () {},
      },
      stubs: {
        RouterLink: RouterLinkStub,
      },
    });
  });

  it('renders a snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('test computed projectUuid()', () => {
    const res = wrapper.vm.projectUuid;
    expect(res).toEqual(project.uuid);
  });

  it('test computed hasNext()', async () => {
    wrapper.setData({
      current: 2,
      total: 4,
    });
    const res = await wrapper.vm.hasNext;
    expect(res).toBeTruthy;
  });

  it('test computed hasPrevious()', async () => {
    wrapper.setData({
      current: 2,
    });
    const res = await wrapper.vm.hasPrevious;
    expect(res).toBeTruthy;
  });

  it('test method startAnimating()', async () => {
    await wrapper.vm.startAnimating();
    expect(wrapper.vm.animating).toBeTruthy;
  });

  it('test method onSelect()', async () => {
    const INDEX = 2;
    const spy = vi.spyOn(wrapper.vm, 'resetTimeout');
    await wrapper.vm.onSelect(INDEX);

    expect(spy).toHaveBeenCalledTimes(1);
    expect(wrapper.vm.current).toEqual(INDEX - 1);
  });

  it('test method onFinishAnimating()', async () => {
    const spy = vi.spyOn(wrapper.vm, 'next');
    await wrapper.vm.onFinishAnimating();

    expect(spy).toHaveBeenCalledTimes(1);
    expect(wrapper.vm.animating).toBeFalsy;
  });
});
