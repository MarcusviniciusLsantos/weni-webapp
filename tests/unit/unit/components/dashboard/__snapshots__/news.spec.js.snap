// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`news.vue > renders a snapshot 1`] = `
<div
  class="weni-news"
  data-v-9a75ab8d=""
>
  <div
    class="weni-news__content"
    data-v-9a75ab8d=""
  >
    <div
      class="weni-news__content__text"
      data-v-9a75ab8d=""
    >
      <h2
        data-v-9a75ab8d=""
      >
        <span
          data-v-9a75ab8d=""
        >
          All in one place
        </span>
      </h2>
      <p
        data-v-9a75ab8d=""
      >
        You can create bots with artificial inteligence and integrate them to your communication systems.
      </p>
    </div>
    <div
      class="weni-news__content__text"
      data-v-9a75ab8d=""
      style="display: none;"
    >
      <h2
        data-v-9a75ab8d=""
      >
        Go to Flows â†’
      </h2>
      <p
        data-v-9a75ab8d=""
      >
        In 
        <b>
          Flows
        </b>
         you can create bots, manage conversation flows and set your custom contact settngs.
      </p>
    </div>
  </div>
  <div
    class="weni-news__controls"
    data-v-9a75ab8d=""
  >
    <div
      class="weni-news__indicator__wrapper"
      data-v-9a75ab8d=""
    >
      
      <div
        class="weni-news__indicator weni-news__indicator--active"
        data-v-9a75ab8d=""
      />
      <div
        class="weni-news__indicator"
        data-v-9a75ab8d=""
      />
      <div
        class="weni-news__indicator"
        data-v-9a75ab8d=""
      />
      <div
        class="weni-news__indicator"
        data-v-9a75ab8d=""
      />
      
    </div>
  </div>
</div>
`;
